runtime/edit-output.cpp

```diff
@@ -1,91 +1,160 @@
-//===-- runtime/numeric-output.h --------------------------------*- C++ -*-===//
+//===-- runtime/edit-output.cpp ---------------------------------*- C++ -*-===//
 //
 // Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
 // See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
 //===----------------------------------------------------------------------===//
 
-#ifndef FORTRAN_RUNTIME_NUMERIC_OUTPUT_H_
-#define FORTRAN_RUNTIME_NUMERIC_OUTPUT_H_
-
-// Output data editing templates implementing the FORMAT data editing
-// descriptors E, EN, ES, EX, D, F, and G for REAL data (and COMPLEX
-// components, I and G for INTEGER, and B/O/Z for both.
-// See subclauses in 13.7.2.3 of Fortran 2018 for the
-// detailed specifications of these descriptors.
-// List-directed output (13.10.4) for numeric types is also done here.
-// Drives the same fast binary-to-decimal formatting templates used
-// in the f18 front-end.
-
-#include "format.h"
-#include "io-stmt.h"
-#include "flang/Decimal/decimal.h"
+#include "edit-output.h"
+#include "flang/Common/uint128.h"
+#include "flang/Common/unsigned-const-division.h"
 
 namespace Fortran::runtime::io {
 
-class IoStatementState;
-
-// I, B, O, Z, and G output editing for INTEGER.
-// edit is const here (and elsewhere in this header) so that one
-// edit descriptor with a repeat factor may safely serve to edit
-// multiple elements of an array.
-bool EditIntegerOutput(IoStatementState &, const DataEdit &, std::int64_t);
-
-// Encapsulates the state of a REAL output conversion.
-class RealOutputEditingBase {
-protected:
-  explicit RealOutputEditingBase(IoStatementState &io) : io_{io} {}
-
-  static bool IsDecimalNumber(const char *p) {
-    if (!p) {
-      return false;
+template<typename INT, typename UINT>
+bool EditIntegerOutput(IoStatementState &io, const DataEdit &edit, INT n) {
+  char buffer[130], *end = &buffer[sizeof buffer], *p = end;
+  bool isNegative{false};
+  if constexpr (std::is_same_v<INT, UINT>) {
+    isNegative = (n >> (8 * sizeof(INT) - 1)) != 0;
```

Comment:
I get this error compiling with clang on macos:
```
../../runtime/edit-output.cpp:20:24: error: conversion from 'unsigned long' to 'common::UnsignedInt128' is ambiguous
    isNegative = (n >> (8 * sizeof(INT) - 1)) != 0;
                       ^~~~~~~~~~~~~~~~~~~~~
```
I think it is due to `std::size_t` being `unsigned long` and `std::uint64_t` being `unsigned long long`.

Reply:
Thanks for checking.  I'm not able to build on osx again yet (or with clang anywhere else yet for that matter).  I'll add a static_cast.  Were there any other errors?

Reply:
That was the only error.